### PR TITLE
Normalize compile-time linear arithmetic ops to a sum of products

### PR DESCRIPTION
**Pull request by mbovel**

### CODE REVIEW

**Comment by mbovel** [comment]
Normalization does not play well with type parameters inference:

```scala
val m: Long = 42
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 2L * m.type // normalized to m.type * 2L
     = mult(2L, m) // A is inferred to be 2L and B to be m.type
```

```
-- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:7:30 
7 |    val z: 2L * m.type = mult(2L, m)
  |                              ^^
  |                              Found:    (2L : Long)
  |                              Required: (m : Long)

longer explanation available when compiling with `-explain`
-- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:7:34 
7 |    val z: 2L * m.type = mult(2L, m)
  |                                  ^
  |                                  Found:    (m : Long)
  |                                  Required: (2L : Long)
```

Note that typing `mult` using dependent parameters instead of type parameters works fine:

```scala
val m: Long = 9
def mult(a: Long, b: Long) = (a * b).asInstanceOf[a.type * b.type]
val z: 2L * m.type = mult(2L, m)
```

**Comment by mbovel** [comment]
This similar case also does not work on `master`:

```scala
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 6L = mult(2L, 3L)
```

```
-- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:8:20 
8 |    val z: 6L = mult(2L, 3L)
  |                ^^^^^^^^^^^^
  |                Found:    Long * Long
  |                Required: (6L : Long)
```

To get more precise types inferred, `Singleton` bounds are needed:

```scala
def mult3[A <: Long & Singleton, B <: Long & Singleton](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 6L = mult3(2L, 3L)
```

**Comment by soronpo** [comment]
> This similar case also does not work on `master`:
> 
> ```scala
> def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
> val z: 6L = mult(2L, 3L)
> ```
> 
> ```
> -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:8:20 
> 8 |    val z: 6L = mult(2L, 3L)
>   |                ^^^^^^^^^^^^
>   |                Found:    Long * Long
>   |                Required: (6L : Long)
> ```
> 
> To get more precise types inferred, `Singleton` bounds are needed:
> 
> ```scala
> def mult3[A <: Long & Singleton, B <: Long & Singleton](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
> val z: 6L = mult3(2L, 3L)
> ```

Yes, scalac is an eager widener.
Try this:
inline def mult(inline a: Long, inline b: Long): a.type * b.type = (a * b).asInstanceOf[a.type * b.type]

**Comment by soronpo** [comment]
I see here that you are adding support for `1 + 1L` with "implicit type conversion". 
It was discussed on the original PR that added compiletime-ops, and IMO shouldn't be directly implemented in the compiler because we can use match types to do this.
```scala
import compiletime.ops.{int, long}
type +[A <: Int | Long, B <: Int | Long] : Int | Long = (A, B) match
  case (Int, Int) => int.+[A, B]
  case (Long, Long) => long.+[A, B]
  case (Int, Long) => long.+[int.ToLong[A], B]
  case (Long, Int) => long.+[A, int.ToLong[B]]
```
The better approach is to add a package `compiletime.ops.all` that combines all the operations, and uses match types where applicable for common operations.

**Comment by mbovel** [comment]
> The better approach is to add a package `compiletime.ops.all` that combines all the operations, and uses match types where applicable for common operations.

Ok, I removed implicit conversions from this PR.

File: compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala [hunk 1]
```diff
@@ -750,7 +750,11 @@ object ProtoTypes {
         else
           newDepTypeVar(ref)
       mt.resultType.substParams(mt, mt.paramRefs.map(replacement))
-    else mt.resultType
+    else
+      // TODO(mbovel): document this.
+      mt.resultType match
+        case at@AppliedType(tycon: TypeRef, args) if defn.isCompiletimeAppliedType(tycon.symbol) => at.underlying
+        case _ => mt.resultType
```
Comment (mbovel) [inline compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:757]: @smarter this is my proposition to fix the problem I mentioned in our last meeting.

Before this change :

```scala
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 6L = mult(2L, 3L) // 6L does not adds constraints on A or B, so there are inferred as Long and Long

// -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:8:20 
// 8 |    val z: 6L = mult(2L, 3L)
//   |                ^^^^^^^^^^^^
//   |                Found:    Long * Long
//   |                Required: (6L : Long)
```

```scala
val m: Long = 42
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: 2L * m.type // normalized to m.type * 2L
     = mult(2L, m) // A is inferred to be 2L and B to be m.type

// -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:7:30 
// 7 |    val z: 2L * m.type = mult(2L, m)
//   |                              ^^
//   |                              Found:    (2L : Long)
//   |                              Required: (m : Long)
// 
// longer explanation available when compiling with `-explain`
// -- [E007] Type Mismatch Error: tests/pos/singleton-ops-long-normalize.scala:7:34 
// 7 |    val z: 2L * m.type = mult(2L, m)
//   |                                  ^
//   |                                  Found:    (m : Long)
//   |                                  Required: (2L : Long)

val m: Long = 42
def mult[A <: Long, B <: Long](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
val z: m.type * 2L // already in normal form
     = mult(2L, m) // A is inferred to be m.type and B to be 2L, so this works
```

After this change, singletons type parameters are never inferred for compile-time operations, so all of the 3 above cases fail with the same message:

```
Found:    Long * Long
Required: …
```

However, typing `mult` with `Singleton` bounds or with dependent parameters works fine before and after for all 3 cases:

```scala
def mult2(a: Long, b: Long) = (a * b).asInstanceOf[a.type * b.type]
def mult3[A <: Long & Singleton, B <: Long & Singleton](a: A, b: B): A * B = (a * b).asInstanceOf[A * B]
```

What do you think? Would that be a more coherent behavior? Or should we not change it?

In both cases tf-dotty would need to be updated to type operations as `mult2` or `mult3`. See https://github.com/MaximeKjaer/tf-dotty/blob/master/modules/compiletime/src/main/scala/io/kjaer/compiletime/dependent.scala#L9. I can't find of a simple solution that would not require changes in tf-dotty.

Comment (smarter) [reply compiler/src/dotty/tools/dotc/typer/ProtoTypes.scala:757]: > Would that be a more coherent behavior? 

No. It adds a special case in inference making it behave differently than for normal operations and the end result seems to only break more things, so I wouldn't do this.

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
test performance with #compiletime please

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
test performance please

**Comment by mbovel** [comment]
test performance please

**Pull request closed without merging**